src/lex.cpp

```diff
@@ -771,14 +777,19 @@ char8* lexer::check_garbage_in_number_literal(char8* input) {
 }
 
 void lexer::parse_binary_number() {
-  QLJS_ASSERT(this->is_binary_digit(this->input_[0]));
   char8* input = this->input_;
 
   while (this->is_binary_digit(*input)) {
     input += 1;
   }
 
-  this->input_ = check_garbage_in_number_literal(input);
+  char8 c = *(input + 1);
+  if (input == this->input_ && (this->is_space(c) || c == u8'\0')) {
+    this->error_reporter_->report(error_no_digits_in_binary_number{
+        source_code_span(input, input)});
+  } else {
+    this->input_ = check_garbage_in_number_literal(input);
+  }
```

Comment:
What about code like this:
```javascript
let x = 0b;
let y = myArray[0b];
```
I think your implementation would fail here. You're only checking for `\0`/EOF and whitespace, not for characters like `;` and `]`. Maybe I'm wrong, though; I didn't analyze this super carefully.

What about an implementation like this:
```suggestion
  this->input_ = check_garbage_in_number_literal(input);
  if (input == this->input_) {
    this->error_reporter_->report(error_no_digits_in_binary_number{
        source_code_span(input, input)});
  }
```

Reply:
Indeed, my code was failing on the cases  you mentioned. so I changed the code as you suggested above and
It's now more generalized than before. But I don't know how to write the test case for multiple multiple expression.
I only checked a few cases by manually running the executable.

Reply:
ok, `check_tokens_with_errors` in `test-lex.cpp` expects to testing ends with [`token_type::end_of_file`](https://github.com/catac1/quick-lint-js/blob/1538d2a80a24032d71d70e17a35e9d2852acbf25/test/test-lex.cpp#L1426) 
so we can't test like `"0b;"` to test `parse_binary_number` because it will ends with `token_type::semicolon`

Reply:
I think there's three options here.

1. Just check errors manually without using `check_tokens_with_errors`.
   Easy but terrible code duplication expected.

2. Write another function which expecting `token_type::semicolon` after the parsing.
   Also easy with less code duplication than  option 1 but still not flexible enough because there may exist 
   different expected ending `token_type` in the future.

3. Add additional parameter `token_type` to the `check_tokens_with_errors` with default parameter `token_type::end_of_file`.  Most flexible one in these options. 

I'm not sure about option 2 and 3 is in the scope of this pull request.

Reply:
check_tokens_with_errors takes a list of tokens, not just a single token. So you can write something like this:

```cpp
check_tokens_with_errors(
    u8"0b;", {token_type::number, token_type::semicolon},
    ...);
```

Reply:
All right. I pushed some more test cases. but I found another failing case  even with this change.
Currently, the `lex` produces `{token_type::number, token_type::identifier}`  for  `"0b_"`  because `lex::parse_binary_number` doesn't consume underscore characters when there's no digit exists in number literals.  

Below cases are unclear to me what is the correct range of the `error_no_digits_in_binary_number` error.

1. ```
   A        B
   0b_11    0b_11
   ^^^      ^^^^^
   ```               

 
2. ```
   A        B        C
   0b_13    0b_13    0b_13
   ^^^      ^^^^     ^^^^^
   ```


3. ```
   A        B         C
   0b__11   0b__11    0b__11          
   ^^^      ^^^^      ^^^^^^
   ```

Reply:
Hmm, that's a good question. For these cases, I think we should create a new error type, `error_number_literal_contains_leading_underscores`. That's out of scope for this PR, I think. Let's file an issue, and accept your current behavior for now.